test: p2p: check disconnect due to lack of desirable service flags

theStack opened the pull request:
This PR adds missing test coverage for disconnecting peers which don't offer the desirable service flags in their VERSION message:
https://github.com/bitcoin/bitcoin/blob/5f3a0574c45477288bc678b15f24940486084576/src/net_processing.cpp#L3384-L3389
This check is relevant for the connection types "outbound-full-relay", "block-relay-only" and "addr-fetch" (see `CNode::ExpectServicesFromConn(...)`). Feeler connections always disconnect, which is also tested here.

In lack of finding a proper file where this test would fit in, I created a new one. Happy to take suggestions there.

theStack committed:
test: p2p: check disconnect due to lack of desirable service flags

delta1 reviewed (approved):
Code Review ACK 358561e54ef84f64b216ca8a4271ad17d91b1d51

Adds a new test, increases code coverage.

furszy reviewed (commented):
See #28170. Obvious concept ACK.
We do have a post-IBD bug on this process.

BrandonOdiwuor reviewed (commented):
Concept ACK 358561e54ef84f64b216ca8a4271ad17d91b1d51

Increasing code coverage

stratospher commented on test/functional/p2p_handshake.py:
```diff
@@ -0,0 +1,46 @@
+        node = self.nodes[0]
+        self.log.info("Check that lacking desired service flags leads to disconnect")
+        for i, conn_type in enumerate(["outbound-full-relay", "block-relay-only", "addr-fetch"]):
+            services = random.choice([0, NODE_NETWORK, NODE_WITNESS])
```
358561e: nit: s/0/`NODE_NONE`

stratospher reviewed (commented):
Concept ACK! i think flaky tests are possible because disconnection could happen before [`p2p_conn.wait_for_connect()`](https://github.com/bitcoin/bitcoin/blob/7699a1aab8fa2a7a602391025f3cfcde29ff2613/test/functional/test_framework/test_node.py#L728) in `add_outbound_connection()`?

theStack committed:
test: p2p: support disconnect waiting for `add_outbound_p2p_connection`

Adds a new boolean parameter `wait_for_disconnect` to the
`add_outbound_p2p_connection` method. If set, the node under
test is checked to disconnect immediately after receiving the
version message (same logic as for feeler connections).

stratospher commented on test/functional/p2p_handshake.py:
```diff
@@ -0,0 +1,49 @@
+
+        self.log.info("Check that feeler connections get disconnected immediately")
+        with node.assert_debug_log(["feeler connection completed"]):
+            self.check_outbound_disconnect(node, i+1, "feeler", 0)
```
ff5f994: nit: `NODE_NONE` here too?

stratospher reviewed (commented):
tested ACK ff5f994.

stratospher commented:
ACK d82eafb173d6bfa98a59e86a845013cc8528b65d.

Empact commented on test/functional/p2p_handshake.py:
```diff
@@ -0,0 +1,49 @@
+        node = self.nodes[0]
+        self.log.info("Check that lacking desired service flags leads to disconnect")
+        for i, conn_type in enumerate(["outbound-full-relay", "block-relay-only", "addr-fetch"]):
+            services = random.choice([NODE_NONE, NODE_NETWORK, NODE_WITNESS])
```
Why not enumerate all combinations of services and connection types here?

epiccurious replied:
ACK this nit. Any reason to not include all here?

Empact replied:
Benefits of enumerating: tests will be comprehensive and ensure that results can be consistently reproduced on every run, instead of being possibly flaky.
Costs: 9 instead of 3 iterations of this case.

Seems to weigh in favor of iteration - imo our tests should tend towards comprehensive and consistent.

cbergqvist replied:
(ACKing nit).

furszy reviewed (commented):
Now that #28170 has been merged, it would be good to include a test case for it.
The node should only connect to `NODE_NETWORK_LIMITED` peers when the local chain is within 24h window from the tip.

epiccurious commented:
> good to include a test case ... only connect to NODE_NETWORK_LIMITED peers when the local chain is within 24h window from the tip

ACK

epiccurious commented:
utACK d82eafb173d6bfa98a59e86a845013cc8528b65d.

fjahr reviewed (commented):
tACK d82eafb173d6bfa98a59e86a845013cc8528b65d

cbergqvist reviewed (commented):
ACK d82eafb.

## ++Verbosity

### Review
Test code is fairly clear.

### Tested (23 Feb)
Ran `./test/functional/p2p_handshake.py` a bunch of times unmodified and observed the different service flags being chosen randomly and tests succeeding.

#### Poked (27 Feb)
Modified `p2p_handshake.py` to always make the outbound connection have `DESIRABLE_SERVICE_FLAGS`. Observed that the outbound node would no longer get disconnected.

theStack committed:
test: p2p: check limited peers desirability (depending on best block depth)

This adds coverage for the logic introduced in PR #28170
("p2p: adaptive connections services flags").

fjahr commented on test/functional/p2p_handshake.py:
```diff
@@ -60,6 +62,13 @@ def run_test(self):
+        self.log.info("Check that limited peers are only desired if the local chain is close to the tip (<24h)")
+        node.setmocktime(int(time.time()) + 25 * 3600)  # tip outside the 24h window, should fail
+        self.test_desirable_service_flags(node, [NODE_NETWORK_LIMITED | NODE_WITNESS], expect_disconnect=True)
+        self.generate(node, 1)  # catch up by mining a block
+        node.setmocktime(int(time.time()) + 23 * 3600)  # tip inside the 24h window, should succeed
```
nit: This is a bit redundant. Either mining the block or changing the mock time alone will make the following test succeed.

davidgumberg replied:
I think it's a bug, since this causes `ApproximateBestBlockDepth()` to return a negative number, we are not really testing the behavior of `if (ApproximateBestBlockDepth() < NODE_NETWORK_LIMITED_ALLOW_CONN_BLOCKS)`

This fixes it:
```suggestion
        self.generate(node, 1)  # catch up by mining a block
        node.bumpmocktime(23 * 3600)  # tip inside the 24h window, should succeed
```

fjahr reviewed (commented):
re-ACK 2f5b69eaeb623a7b0306e7f41fd2d28741f131c0

cbergqvist reviewed (approved):
ACK 2f5b69e!

Tried to make sense of `git range-diff d82eafb~2..d82eafb 2f5b69e~3..2f5b69e` but the GitHub Files changed view did help.

Ran
```sh
./test/functional/p2p_handshake.py
```
and
```sh
./test/functional/p2p_handshake.py --v2transport
```
a couple of times each.

marcofleon reviewed (commented):
Good work, ACK 2f5b69eaeb623a7b0306e7f41fd2d28741f131c0. Read over the changes, built the PR, ran all functional tests and then individually ran `p2p_handshake.py` and `p2p_handshake.py --v2transport`. I also inverted some of the logic of `expect_disconnect` throughout the test. Everything passed/failed as expected.

The test is designed well and is easily extensible, so I'd say current test coverage is sufficient for merging.

mzumsande commented on test/functional/test_framework/test_node.py:
```diff
@@ -726,7 +726,7 @@ def add_p2p_connection(self, p2p_conn, *, wait_for_verack=True, send_version=Tru
         return p2p_conn
 
-    def add_outbound_p2p_connection(self, p2p_conn, *, wait_for_verack=True, p2p_idx, connection_type="outbound-full-relay", supports_v2_p2p=None, advertise_v2_p2p=None, **kwargs):
+    def add_outbound_p2p_connection(self, p2p_conn, *, wait_for_verack=True, wait_for_disconnect=False, p2p_idx, connection_type="outbound-full-relay", supports_v2_p2p=None, advertise_v2_p2p=None, **kwargs):
```
nit: Maybe it would be more flexible to just have a `expect_success` variable here, and let the calling test deal with the waiting for the disconnect. That way, in future tests it could also be used to test for disconnections that happen before sending the version message (e.g. during the v2 handshake, similar to what #29431 does).

mzumsande reviewed (commented):
Code Review ACK 2f5b69eaeb623a7b0306e7f41fd2d28741f131c0

fjahr commented:
re-ACK 49fbd57eebe60c3a36459df7b1545c2c477b2e68b

davidgumberg commented:
Tested ACK https://github.com/bitcoin/bitcoin/pull/29279/commits/49fbd57eebe60c3a36459df7b1545c2c477b2e68

Modified values and logic in `HasAllDesirableServiceFlags` and `GetDesirableServiceFlags` and verified that the tests added here complain.

It might be nice to have a test to ensure that we won't evict manual or inbound connections for having bad service flags, but as far as I can tell, that would require more refactoring of the test framework.

One question that still lingers for me after reviewing this is why the limit for `NETWORK_LIMITED` being desirable  added in https://github.com/bitcoin/bitcoin/pull/28170 set  `NODE_NETWORK_LIMITED_ALLOW_CONN_BLOCKS = 144;` instead of `288` which is how many blocks BIP 159 requires `NETWORK_LIMITED` nodes to store.

cbergqvist reviewed (approved):
ACK 49fbd57.

Still have some things to learn about git diffing it seems. I revoke my review of 2f5b69e and re-raise with this one now that I have absorbed a rough idea of `NODE_NETWORK_LIMITED` and `ApproximateBestBlockDepth()`. Cheers @davidgumberg!


https://github.com/bitcoin/bitcoin/blob/49fbd57eebe60c3a36459df7b1545c2c477b2e68/test/functional/p2p_handshake.py#L66-L69

nit: Would feel more comfortable if the test moved forwards in time instead of backwards, setting the node's mocked hours to 23 *before* 25. I re-tested `p2p_handshake.py` with and without `--v2transport`, both with backwards time travel as currently implemented, and switching around the lines for forwards time travel (also tried a version using `bumpmocktime()` 2 hours which also worked, but may be less clear).

furszy reviewed (commented):
> One question that still lingers for me after reviewing this is why the limit for NETWORK_LIMITED being desirable (added in https://github.com/bitcoin/bitcoin/pull/28170) is: NODE_NETWORK_LIMITED_ALLOW_CONN_BLOCKS = 144; instead of 288 which is how many blocks BIP 159 requires NETWORK_LIMITED nodes to store.

@davidgumberg start from this comment https://github.com/bitcoin/bitcoin/pull/28170#pullrequestreview-1825332930. We had the same conversation there.

stratospher commented on test/functional/p2p_handshake.py:
```diff
@@ -0,0 +1,69 @@
+
+
+# usual desirable service flags for outbound non-pruned peers
+DESIRABLE_SERVICE_FLAGS = NODE_NETWORK | NODE_WITNESS
```
b6c7ae6: shouldn't we use `NODE_NETWORK_LIMITED | NODE_WITNESS` for pruned peers debug log in 49fbd57e.

davidgumberg commented:
reACK https://github.com/bitcoin/bitcoin/commit/2f23987849758537f76df7374d85a7e87b578b61

Looks good, retested by modifying `HasAllDesirableServiceFlags` and `ExpectServicesFromConn`

fjahr commented on test/functional/p2p_handshake.py:
```diff
@@ -47,18 +53,30 @@ def test_desirable_service_flags(self, node, service_flag_tests, expect_disconne
+        self.test_desirable_service_flags(node, [NODE_NETWORK_LIMITED | NODE_WITNESS],
+                                          DESIRABLE_SERVICE_FLAGS_FULL, expect_disconnect=True)
+        node.setmocktime(int(time.time()) + 23 * 3600)  # tip inside the 24h window, should succeed
+        self.test_desirable_service_flags(node, [NODE_NETWORK_LIMITED | NODE_WITNESS],
```
nit: The `DESIRABLE_SERVICE_FLAGS_*` above are the same as the combinations passed as `service_flag_tests` so that variable could have been used in both places IMO (with a slight rename maybe). But this isn't critical.

fjahr reviewed (commented):
re-utACK 2f23987849758537f76df7374d85a7e87b578b61

(CI failure can be ignored)

cbergqvist reviewed (approved):
re ACK 2f23987849758537f76df7374d85a7e87b578b61

Built & re-ran `test/functional/test_runner.py --cachedir=/mnt/tmp/cache --tmpdir=/mnt/tmp`.

stratospher reviewed (commented):
tested ACK 2f23987. 🚀
